### ISSUE TITLE
splits stealing from slow worker

### ISSUE THREAD

**Issue by 51yu** [trinodb/trino#7380]
Time cost of large hive table scanning (persist in s3) is upper bounded by the worker with least I/O, in multi-tenant (cloud environment) worker I/O is not always guaranteed (noisy neighbor, shared hardware resource - network bandwidth / CPU etc).
Query manager (coordinator) should be aware of such slow worker during runtime (on demand post-planning), so that peer (healthy) workers should be able to steal splits (ex: s3 file reading) from slow worker to unblock stage making progress.

**Comment by sopel39** [comment]
Workers have limited split queue length to prevent just that. How much lag there is in slow worker compared to fast workers?

**Comment by 51yu** [comment]
> Workers have limited split queue length to prevent just that. How much lag there is in slow worker compared to fast workers?

tracking the webui dashboard, I found splits on healthy workers finished 8 splits assigned within same stage (splits are evenly distributed across workers) while slow worker almost make no progress for a single split among 8. And network throughput of a healthy worker is 30x comparing the slow one.

**Comment by findepi** [comment]
I would recommend  focusing energy on identifying "slow workers" and then being able to remove them from the cluster.
that would probably require administrator's  action (human or automated) to shut them down and de-provision

**Comment by 51yu** [comment]
> I would recommend focusing energy on identifying "slow workers" and then being able to remove them from the cluster.
> that would probably require administrator's action (human or automated) to shut them down and de-provision

good point - quarantine slow worker, I am wondering whether we could automate this through control plane (coordinator) instead of manual operations

in-long term, I am happy to see dynamic planning happen between workers (stealing splits) - slow worker is transient

**Comment by 51yu** [comment]
or trino community provide an API endpoint at coordinator, so that admin/operator can quarantine slow worker from worker pool with minimum effort (a HTTP request) - operation efficiency @findepi
